components/layout_thread/dom_wrapper.rs

```diff
@@ -411,7 +479,11 @@ impl<'le> TElement for ServoLayoutElement<'le> {
     }
 
     fn traversal_children(&self) -> LayoutIterator<Self::TraversalChildrenIterator> {
-        LayoutIterator(self.as_node().dom_children())
+        LayoutIterator(if let Some(shadow) = self.shadow_root() {
```

Comment:
This is ok for now. It'd be good to introduce the concept of flat tree in a more formal way, but...